src/asyncresults.jl

```diff
@@ -0,0 +1,242 @@
+    log_function = throw_error ? error : warn
+
+    if success
+        async_result.result_task = consume(async_result; throw_error=throw_error)
+    else
+        msg = error_message(async_result.jl_conn)
+        _ensure_unlocked(async_result)
+        log_function(LOGGER, msg)
+    end
+
+    return async_result
+end
+
+"""
+    consume(async_result::AsyncResult; throw_error=true) -> Task
+
+Run a task which executes the query in `async_result` and waits for results.
+
+This implements the loop described in the PostgreSQL documentation for
+[Asynchronous Command Processing](https://www.postgresql.org/docs/10/libpq-async.html).
+
+The `throw_error` option only determines whether to throw errors when handling the new
+[`Result`](@ref)s; the `Task` may error for other reasons related to processing the
+asynchronous loop.
+
+The result returned from the `Task` will be the [`Result`](@ref) of the last query run (the
+only query if using parameters).
+Any errors produced by the queries will be thrown together in a `CompositeException` by
+`@sync`.
+"""
+function consume(async_result::AsyncResult; throw_error=true)
+    @async begin
+        try
+            debug(LOGGER, "getting the socket fd")
+            pqfd = socket(async_result.jl_conn)
+            try
+                debug(LOGGER, "making an FDWatcher")
+                watcher = FDWatcher(pqfd, true, false)  # readable, not writeable
+                try
+                    last_result = @sync begin
+                        result = nothing
+                        debug(LOGGER, "beginning the waiting loop")
+                        while true
+                            debug(LOGGER, "waiting on the fd")
+                            wait(watcher)
+                            debug(LOGGER, "consuming input")
+                            success = libpq_c.PQconsumeInput(async_result.jl_conn.conn) == 1
+                            !success && error(error_message(async_result.jl_conn))
+                            debug(LOGGER, "checking isbusy")
+                            while libpq_c.PQisBusy(async_result.jl_conn.conn) == 0
+                                debug(LOGGER, "getting a result")
+                                result_ptr = libpq_c.PQgetResult(async_result.jl_conn.conn)
+                                result_ptr == C_NULL && @goto finished
+                                debug(LOGGER, "handling the non-null result")
+                                result = @async handle_result(Result(
+                                    result_ptr, async_result.jl_conn;
+                                    async_result.result_kwargs[]...
+                                ); throw_error=throw_error)
+                            end
+                        end
+                        @label finished
+                        debug(LOGGER, "finished the sync block")
+                        result
+                    end
+
+                    return fetch(last_result)
```

Comment:
So we are actually returning a `Future` which is itself a fetch of a nother `Future` (the `handle_result(Result` operation).
But the inner `Future` is already `ready`, because it was un in a `@sync` block.
Which means the time for this one to be ready is the time taken to do the `fetch` operation,
here.

So we are releasing our lock on the connection
after the `last_result` has been `handle`ed, but before it has been `fetch`ed.

Is this:
A) Where we want out asynconicity
B.) Where we want our locks to be released

Reply:
They are `Task`s not `Future`s, to be precise. In the success case this is not a significant distinction, but in the error case it makes a difference. When you fetch a `Task` which threw an error, that error is thrown. So if we didn't call `fetch`, `consume` would return a `Task` for which `fetch` would always return another `Task`. But since we call `fetch`, `consume` returns a `Task` for which fetch will either return a `Result` or throw an error. That's the behaviour we want.

I'm don't think I understand (A) but maybe I answered it?